src/native/libs/System.Security.Cryptography.Native/pal_evp_pkey.c

```diff
@@ -3,91 +3,123 @@
+        assert(handle->libCtx != NULL);
 
-    EVP_PKEY* newKey = EVP_PKEY_new();
-
-    if (newKey == NULL)
-    {
-        return NULL;
+        OSSL_PROVIDER_unload(handle->prov);
+        OSSL_LIB_CTX_free(handle->libCtx);
+        free(handle);
     }
-
-    bool success = true;
-
-    if (currentAlgId == EVP_PKEY_RSA)
-    {
-        const RSA* rsa = EVP_PKEY_get0_RSA(currentKey);
-
-#pragma clang diagnostic push
-#pragma clang diagnostic ignored "-Wcast-qual"
-        if (rsa == NULL || !EVP_PKEY_set1_RSA(newKey, (RSA*)rsa))
+}
 #pragma clang diagnostic pop
-        {
-            success = false;
-        }
-    }
-    else
-    {
-        ERR_put_error(ERR_LIB_EVP, 0, EVP_R_UNSUPPORTED_ALGORITHM, __FILE__, __LINE__);
-        success = false;
-    }
 
-    if (!success)
-    {
-        EVP_PKEY_free(newKey);
-        newKey = NULL;
-    }
+#endif
 
-    return newKey;
+EVP_PKEY* CryptoNative_EvpPkeyCreate(void)
+{
+    ERR_clear_error();
+    return EVP_PKEY_new();
 }
 
-void CryptoNative_EvpPkeyDestroy(EVP_PKEY* pkey)
+void CryptoNative_EvpPkeyDestroy(EVP_PKEY* pkey, void* extraHandle)
 {
     if (pkey != NULL)
     {
         EVP_PKEY_free(pkey);
     }
+
+#ifdef NEED_OPENSSL_3_0
+    if (extraHandle != NULL)
+    {
+        EvpPKeyExtraHandle* extra = (EvpPKeyExtraHandle*)extraHandle;
+        CryptoNative_EvpPkeyExtraHandleDestroy(extra);
+    }
+#else
+    (void)extraHandle;
+    assert(extraHandle == NULL);
+#endif
 }
 
-int32_t CryptoNative_EvpPKeySize(EVP_PKEY* pkey)
+int32_t CryptoNative_EvpPKeyBits(EVP_PKEY* pkey)
 {
     // This function is not expected to populate the error queue with
     // any errors, but it's technically possible that an external
     // ENGINE or OSSL_PROVIDER populate the queue in their implementation,
     // but the calling code does not check for one.
     assert(pkey != NULL);
-    return EVP_PKEY_get_size(pkey);
+    return EVP_PKEY_get_bits(pkey);
 }
 
-int32_t CryptoNative_UpRefEvpPkey(EVP_PKEY* pkey)
+#pragma clang diagnostic push
+// There's no way to specify explicit memory ordering for increment/decrement with C atomics.
+#pragma clang diagnostic ignored "-Watomic-implicit-seq-cst"
+int32_t CryptoNative_UpRefEvpPkey(EVP_PKEY* pkey, void* extraHandle)
 {
     if (!pkey)
     {
         return 0;
     }
 
+#ifdef NEED_OPENSSL_3_0
+    EvpPKeyExtraHandle* extra = (EvpPKeyExtraHandle*)extraHandle;
+
+    if (extra != NULL)
+    {
+        extra->refCount++;
+    }
+#else
+    (void)extraHandle;
+    assert(extraHandle == NULL);
+#endif
+
     // No error queue impact.
     return EVP_PKEY_up_ref(pkey);
 }
+#pragma clang diagnostic pop
+
+int32_t CryptoNative_EvpPKeyType(EVP_PKEY* key)
+{
+    int32_t base_id = EVP_PKEY_get_base_id(key);
+    switch (base_id) {
+    case EVP_PKEY_RSA:
+    case EVP_PKEY_EC:
+    case EVP_PKEY_DSA:
+        return base_id;
+    case EVP_PKEY_RSA_PSS:
+        return EVP_PKEY_RSA;
+    case EVP_PKEY_ED448:
+    case EVP_PKEY_ED25519:
+        return EVP_PKEY_EC;
```

Comment:
Should probably just treat EdDSA and X25519 as unknown for now. They are not supported.

Reply:
I based this on one of the samples on OpenSSL but I don't think anything actually returned this during my testing so will remove

Reply:
I agree with Kevin.  Until we have a scenario where we need to work with ED448 or ED25519 keys, don't put them in this table.